### PR TITLE
Add validation for Endpoint

### PR DESCRIPTION
**Pull request by gmarek**
First version of solution for #2572. Few details need to be discussed.

### CODE REVIEW

**Comment by kazegusuri** [comment]
Do you have a plan to port endpoints registry to generic etcd storage? I have worked on porting endpoints, but not touched validations.

**Comment by gmarek** [comment]
I guess I do. If you have PR ready-ish, we can try to sync up.

**Comment by kazegusuri** [comment]
Then I stay on the work. It needs some more time, so I'll sync.

**Comment by gmarek** [comment]
I assume that #5731 will be merged sometime soon. I'll update this PR to handle this.

**Comment by gmarek** [comment]
Updated. @thockin can you please take a look at this PR, as you're working on #5731? I'll add tests and fix this when your PR is merged.

**Comment by thockin** [comment]
Whoah, I'll take a look ASAP.  Please nobody merge this until PRs are in sync

**Comment by gmarek** [comment]
@thockin: I saw that your PR was closed, but there wasn't any reason mentioned. Can you please tell if I should rewrite this one to the "old" (current) types, or is some bigger Endpoint rewrite happening?

**Comment by bgrant0607** [comment]
@gmarek You need to await the "real" PR #5939 to be merged.

**Comment by thockin** [comment]
My big PR is just about ready to commit.  I added a stub for validation into which you can pretty much glue this :)

**Comment by gmarek** [comment]
Thanks Tim!

**Comment by gmarek** [comment]
After moving endpoints to generic etcd and Tims change this PR became really small. But still PTAL

**Comment by gmarek** [comment]
PTAL

**Comment by gmarek** [comment]
@thockin

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -1252,26 +1252,56 @@ func validateEndpointSubsets(subsets []api.EndpointSubset) errs.ValidationErrorL
...
 		}
 		if len(ss.Ports) == 0 {
-			ssErrs = append(ssErrs, errs.NewFieldRequired("ports"))
+			allErrs = append(allErrs, errs.NewFieldRequired("ports"))
+		}
+		for addr := range ss.Addresses {
+			allErrs = append(allErrs, validateEndpointAddress(&ss.Addresses[addr])...)
 		}
-		// TODO: validate each address and port
-		allErrs = append(allErrs, ssErrs.PrefixIndex(i)...)
```
Comment (thockin) [inline pkg/api/validation/validation.go]: I think you've now lost the index - what does the error for this come out as?  It should come out as (for example) "subsets[0].ports[3].name"

File: pkg/api/validation/validation.go [hunk 2]
```diff
@@ -1252,26 +1252,56 @@ func validateEndpointSubsets(subsets []api.EndpointSubset) errs.ValidationErrorL
...
+		}
+	}
+
+	return allErrs
+}
+
+func validateEndpointAddress(address *api.EndpointAddress) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+	if !util.IsValidIPv4(address.IP) {
+		allErrs = append(allErrs, errs.NewFieldInvalid("IP", address.IP, "is invalid IPv4 address"))
```
Comment (thockin) [inline pkg/api/validation/validation.go]: field names should be JSON style - "ip", "name" - start lower-case

File: pkg/api/validation/validation.go [hunk 3]
```diff
@@ -1252,26 +1252,56 @@ func validateEndpointSubsets(subsets []api.EndpointSubset) errs.ValidationErrorL
...
 	}
+	return allErrs
+}
 
+func validateEndpointPort(port *api.EndpointPort, requireName bool) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+	if requireName && port.Name == "" {
+		allErrs = append(allErrs, errs.NewFieldRequired("Name"))
+		return allErrs
```
Comment (thockin) [inline pkg/api/validation/validation.go]: don't short-circuit - we want to report as many errors as we can in a single pass.

File: pkg/api/validation/validation.go (hunk 2 above)
Comment (gmarek) [inline pkg/api/validation/validation.go]: Done

File: pkg/api/validation/validation.go (hunk 3 above)
Comment (gmarek) [inline pkg/api/validation/validation.go]: Yeah - sorry, I thought I removed this, but it must have been in some uncommited change.

File: pkg/api/validation/validation.go (hunk 1 above)
Comment (gmarek) [inline pkg/api/validation/validation.go]: Done.

File: pkg/api/validation/validation.go [hunk 4]
```diff
@@ -1276,17 +1276,51 @@ func validateEndpointSubsets(subsets []api.EndpointSubset) errs.ValidationErrorL
 		if len(ss.Ports) == 0 {
 			ssErrs = append(ssErrs, errs.NewFieldRequired("ports"))
 		}
-		// TODO: validate each address and port
+		for addr := range ss.Addresses {
+			ssErrs = append(ssErrs, validateEndpointAddress(&ss.Addresses[addr])...)
```
Comment (thockin) [inline pkg/api/validation/validation.go]: This still looks like it is going to come out with the wrong field name.  It's pretty ugly, but you have to do something like PrefixIndex(addr).Prefix("addresses") - wee how other array validations are handled.  We're not always consistent, but we should try to be.

File: pkg/api/validation/validation.go [hunk 5]
```diff
@@ -1276,17 +1276,51 @@ func validateEndpointSubsets(subsets []api.EndpointSubset) errs.ValidationErrorL
...
 		allErrs = append(allErrs, ssErrs.PrefixIndex(i)...)
 	}
 
 	return allErrs
 }
 
+func validateEndpointAddress(address *api.EndpointAddress) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+	if !util.IsValidIPv4(address.IP) {
+		allErrs = append(allErrs, errs.NewFieldInvalid("ip", address.IP, "is invalid IPv4 address"))
```
Comment (thockin) [inline pkg/api/validation/validation.go]: I would drop the "is" in the error - it's not generally formatted as a sentence.

Comment (gmarek) [inline pkg/api/validation/validation.go]: Done.

File: pkg/api/validation/validation.go (hunk 4 above)
Comment (gmarek) [inline pkg/api/validation/validation.go]: Done.

**Comment by gmarek** [comment]
PTAL

**Comment by thockin** [comment]
LGTM

**Pull request merged**